# Tarefa 1: Classificação de Variáveis Estatísticas
No dicionário de dados da PRF, encontramos diversas informações sobre as ocorrências.
Considere o seguinte subconjunto de colunas presentes no dataset:

1. dia_semana: Dia da semana em que ocorreu o acidente (ex: “segunda-feira”, “terça-feira”).

2. fase_dia: Luminosidade no momento do acidente (ex: “Pleno dia”, “Anoitecer”, “Plena noite”).

3. idade: Idade da vítima envolvida no acidente.

4. veiculos: Quantidade total de veículos envolvidos na ocorrência.

Classifique cada uma das variáveis acima em Qualitativa (Nominal ou Ordinal) ou Quantitativa
(Discreta ou Contínua). Justifique computacional e estatisticamente sua escolha.

In [3]:
# Importação da biblioteca
import pandas as pd

# Carregamento do dataset acidentes2018.csv
df = pd.read_csv("acidentes2018.csv", encoding="ISO-8859-1", sep=";")

# Exibição das informações do DataFrame e das colunas a serem analisadas
df.info()

print("\n", df["dia_semana"].head(1))
print("\n", df["fase_dia"].head(1))
print("\n", df["idade"].head(1))
print("\n", df["tipo_veiculo"].head(1))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180061 entries, 0 to 180060
Data columns (total 35 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   id                      180061 non-null  float64
 1   pesid                   180061 non-null  int64  
 2   data_inversa            180061 non-null  object 
 3   dia_semana              180061 non-null  object 
 4   horario                 180061 non-null  object 
 5   uf                      180061 non-null  object 
 6   br                      180061 non-null  int64  
 7   km                      180061 non-null  object 
 8   municipio               180061 non-null  object 
 9   causa_acidente          180061 non-null  object 
 10  tipo_acidente           180006 non-null  object 
 11  classificacao_acidente  180058 non-null  object 
 12  fase_dia                180061 non-null  object 
 13  sentido_via             180061 non-null  object 
 14  condicao_metereologi

1 - dia_semana = Qualitativa nominal -> Qualitativa (ou categórica) por não ser uma variável que armazena dados numéricos e nominal porque seus valores não seguem uma ordem hierárquica;

2 - fase_dia = Qualitativa ordinal -> Qualitativa (ou categórica) por não ser uma variável que armazena dados numéricos e ordinal porque seus valores seguem uma ordem hierárquica;

3 - idade = Quantitativa contínua -> Quantitativa porque a variável armazena um valor numérico e contínua porque o valor armazenado não é fracionário;

4 - veiculos = Quantitativa contínua -> Quantitativa porque a variável armazena um valor numérico e contínua porque o valor armazenado não é fracionário.

# Tarefa 2: Codificação de Variáveis Categóricas (Encoding)

1. Algoritmos de aprendizado de máquina frequentemente não lidam bem com strings brutas,
exigindo a transformação das variáveis categóricas para o formato numérico.

2. Selecione a variável qualitativa ordinal fase_dia, que contém os níveis de luminosidade.
Aplique a técnica de Label Encoding (Codificação de Rótulos), atribuindo valores inteiros que
respeitem a gradação natural de luz do dia. Apresente o dicionário de mapeamento gerado
(Chave: String -> Valor: Inteiro) e mostre as primeiras 5 linhas do dataset após a transformação.

In [4]:
# Mudança dos valores da coluna fase_dia para inteiro
df.loc[df["fase_dia"] == "Amanhecer", "fase_dia"] = 1
df.loc[df["fase_dia"] == "Pleno dia", "fase_dia"] = 2
df.loc[df["fase_dia"] == "Anoitecer", "fase_dia"] = 3
df.loc[df["fase_dia"] == "Plena Noite", "fase_dia"] = 4

df["fase_dia"].head()

,fase_dia
0,2
1,2
2,2
3,2
4,1


# Tarefa 3: Categorização de Variáveis Numéricas (Binning)

1. A categorização (ou binning) é útil para reduzir o ruído de variáveis contínuas ou criar perfis
demográficos, transformando variáveis quantitativas em qualitativas ordinais.

2. Utilizando a variável quantitativa idade, crie uma nova coluna no seu DataFrame chamada
faixa_etaria. Utilize a seguinte regra para a discretização (intervalos fechados à esquerda e
abertos à direita):

- [0, 18): Menores de idade
- [18, 30): Jovens
- [30, 60): Adultos
- [60, 120): Idosos

In [5]:
# Criação da nova coluna com binning
bins = [0, 18, 30, 60, 120]
labels = ["Menor de Idade", "Jovem", "Adulto", "Idoso"]
df["faixa_etaria"] = pd.cut(df["idade"], bins=bins, labels=labels, right=False)

print(df["idade"].head())
print("\n", df["faixa_etaria"].head())

0    52
1    78
2     0
3    60
4    66
Name: idade, dtype: int64

 0            Adulto
1             Idoso
2    Menor de Idade
3             Idoso
4             Idoso
Name: faixa_etaria, dtype: category
Categories (4, object): ['Menor de Idade' < 'Jovem' < 'Adulto' < 'Idoso']


#Tarefa 4: Distribuição de Frequências para Variáveis Categóricas

Após a criação da variável categórica faixa_etaria, precisamos entender o perfil das vítimas
dos acidentes. Suponha que, ao extrair uma amostra do dataset, você obteve a seguinte
contagem para a nova variável:
- Menores de idade: 150 ocorrências
- Jovens: 320 ocorrências
- Adultos: 430 ocorrências
- Idosos: 100 ocorrências

  a) Construa a tabela de distribuição de frequências completa para a variável faixa_etaria.
  
  Sua tabela deve conter as seguintes colunas:

  - Frequência Absoluta (fi)
  - Frequência Relativa (fr) expressa em porcentagem
  - Frequência Acumulada (Fi)
  - Frequência Relativa Acumulada (Fr) expressa em porcentagem

  Um exemplo de como a tabela deve ficar é

  | Faixa Etária | *fi* | *fr* (%) | Fi | Fr (%) |
  | :-------- | :--: | :------ | :-- | :---- |
  | Menores de idade | 150 | 15% | 150 | 15% |
  | Jovens | 320 | 32% | 470 | 47% |
  | Adultos | 430 | 43% | 900 | 90% |
  | Idosos | 100 | 10% | 1000 | 100% |

  b) Com base exclusivamente na tabela gerada no item (a), elabore um parágrafo interpretativo. Responda explicitamente: Qual é o grupo de maior risco de envolvimento em acidentes?

  Qual é a porcentagem total de vítimas que possuem menos de 60 anos? Como essa informação
  de frequência acumulada pode ser útil para políticas públicas de prevenção de acidentes da PRF?
  

In [37]:
# Quantidade total de elementos em faixa_etaria
total = df['faixa_etaria'].value_counts().sum()
tam = len(df['faixa_etaria'].value_counts())

# Lista para armazenamento dos dados
fi = []
fr = []
Fr = []
Fi = []

freqAcu = 0
freqRelAcu = 0
variaveis = ["Menor de Idade", "Jovem", "Adulto", "Idoso"]
for x in range(tam):
  # Calcula a frequência absoluta
  totVar = (df["faixa_etaria"] == f"{variaveis[x]}").sum()
  fi.append(totVar)

  # Calcula a frequência acumulada
  freqAcu += fi[x]

  # Calcula a frequência relativa
  freqRel = (fi[x] * 100) / total

  # Calcula a frequência relativa acumulada
  freqRelAcu += freqRel.round(0)

  # Adiciona os dados nas respectivas listas
  fr.append(f"{freqRel:.0f}%")
  Fr.append(f"{freqRelAcu:.0f}%")
  Fi.append(freqAcu)

# Criação de um DataFrame com os dados
table = {
    "Faixa Etária": variaveis,
    "fi": fi,
    "fr": fr,
    "Fi": Fi,
    "Fr": Fr
}

tabela = pd.DataFrame(table)
tabela.head()

,Faixa Etária,fi,fr,Fi,Fr
0,Menor de Idade,38180,21%,38180,21%
1,Jovem,38735,22%,76915,43%
2,Adulto,88889,49%,165804,92%
3,Idoso,13945,8%,179749,100%


A partir da análise da tabela é possível estabelecer que:

- Adultos são o grupo de maior risco em acidentes, tanto em termos absolutos quanto relativos. Pessoas nessa faixa etária ocupam 49% dos acidentes documentados, somando quase 90 mil casos apenas em 2018

- A porcentagem total de vítimas que possuem menos de 60 anos (menores de idade + jovens + adultos) totaliza 92% dos acidentes documentados. Essa informação é importante pois, a partir dela, políticas públicas podem ser tomadas para conscientização e prevenção de acidentes focadas em pessoas abaixo dos 60 anos. Um exemplo seria a conscientização do seguimento das leis de trânsito, treinamento para condutores de veículos e aumento da fiscalização focado em pessoas que se encaixam nessa categoria